// Mocp/lib/Mocp.js
'use strict';

var child_process = require('child_process');

var Mocp = function() {
	console.log('Mocp!');
	
	// https://github.com/andreas-marschke/misc-tools/blob/master/lisp/mocp.el
	
	var _format = '\"%state %song - %album - %artist - %ct of %tt - %r %b\"';
	
	/*
	%state     State
	                       %file      File
	                       %title     Title
	                       %artist    Artist
	                       %song      SongTitle
	                       %album     Album
	                       %tt        TotalTime
	                       %tl        TimeLeft
	                       %ts        TotalSec
	                       %ct        CurrentTime
	                       %cs        CurrentSec
	                       %b         Bitrate
	                       %r         Rate
	*/
	
	this.sendCommand = function(cmd) {
		var buffer = child_process.execSync(cmd);
		console.log(buffer.toString('utf-8'));
	};
	
	this.unpause = function() {
		this.sendCommand('mocp --unpause');
	};
	
	this.pause = function() {
		this.sendCommand('mocp --pause');
	};
	
	this.togglePause = function() {
		this.sendCommand('mocp --toggle-pause');
	};
	
	this.previous = function() {
		this.sendCommand('mocp --previous --format ' + _format);
	};
	
	this.next = function() {
		this.sendCommand('mocp --next --format ' + _format);
	};
	
	this.volumeUp = function() {
		this.sendCommand('mocp -v +5');
	};
	
	this.volumeDown = function() {
		this.sendCommand('mocp -v -5');
	};
	
	this.whatsPlaying = function() {
		this.sendCommand('mocp --format \"%song - %album - %artist\"');
	};
};

module.exports = new Mocp();
