// node-mplayer/lib/node-mplayer.js
var cp = require('child_process'),
    events = require('events'),
    fs = require('fs'),
    readline = require('readline'),
    spawn = cp.spawn;

var MPlayer = function(file, options) {

	var _self = this,
		_file,
		_player;
	
	events.EventEmitter.call(this);
		
	this.play = function(file, options) {
		_file = file;
		var args = ['-slave', _file]; // var args = ['-slave', '-quiet', _file];
		if (options!== undefined && options.volume !== undefined) {
			args.push('-volume',  options.volume);
		}
		_player = spawn('mplayer', args);
		_player.stdout.on('data', function(buffer) { console.log(buffer.toString()); } );
	}
	
	this.stop = function() {
    	if (_player !== null) {
        	_player.stdin.write('stop\n');
    	}
    };
    
    this.mute = function() {
    	if (_player !== null) {
        	_player.stdin.write('mute\n');
    	}
    };
    
    this.pause = function() {
    	if (_player !== null) {
        	_player.stdin.write('pause\n');
    	}
    };
	
	this.setVolume = function(volume) {
    	if (_player !== null) {
        	_player.stdin.write('volume ' + volume +  ' 1\n');
    	}
    };
    
	this.setSpeed = function(speed) {
    	if (_player !== null) {
        	_player.stdin.write('speed_set ' + speed + '\n');
    	}
    };
   
};

function Mplayer(path) {
	
	var _self = this;
	
    this.childProc = null;
    this.file = "";
    this.rl = null;

    if(typeof path !== 'undefined')
        this.setFile(path);

    events.EventEmitter.call(this);
	
	/*
    cp.exec('mplayer', function(err, stdout, stdin){
        if(err)
            throw new Error("Mplayer encountered an error or isn't installed.");
    });
    */
};

Mplayer.prototype.__proto__ = events.EventEmitter.prototype;

Mplayer.prototype.play = function(opts) {
    if (this.file !== null) {
        var args = ['-slave', '-quiet', this.file];

        this.childProc = spawn('mplayer', args);

        if (typeof opts !== 'undefined'){
            if (typeof opts.volume !== 'undefined') {
                this.setVolume(opts.volume);
            }

            if (typeof opts.loop !== 'undefined') {
                this.setLoop(opts.loop);
            }
        }

        this.childProc.on('error', function(error){
            _self.emit('error');
        });

        this.childProc.on('exit', function(code, sig) {
			console.log('EXIT');
            if (code === 0 && sig === null)
                _self.emit('end');
        });

		this.childProc.stdout.on('data', function(data) {
			_self.emit('data', data);
        });

        this.rl = readline.createInterface({
            input: this.childProc.stdout,
            output: this.childProc.stdin
        });
    }
};

Mplayer.prototype.stop = function() {
    if (this.childProc !== null){
        this.childProc.stdin.write('stop\n');
    }
};

Mplayer.prototype.pause = function() {
    if (this.childProc !== null) {
        this.childProc.stdin.write('pause\n');
    }
};

Mplayer.prototype.mute = function() {
    if (this.childProc !== null) {
        this.childProc.stdin.write('mute\n');
    }
};

Mplayer.prototype.setVolume = function(volume) {
    if (this.childProc !== null) {
        this.childProc.stdin.write('volume ' + volume + ' 1\n');
    }
};

Mplayer.prototype.seek = function(sec) {
    if (this.childProc !== null) {
        this.childProc.stdin.write('seek ' + sec + ' 2\n');
    }
};

Mplayer.prototype.setLoop = function(times) {
    if (this.childProc !== null) {
        this.childProc.stdin.write('loop ' + times + '\n');
    }
};

Mplayer.prototype.setSpeed = function(speed) {
    if (this.childProc !== null) {
        this.childProc.stdin.write('speed_set ' + speed + '\n');
    }
};

Mplayer.prototype.setFile = function(path) {
    if (fs.existsSync(path)) {
        this.file = path;
    } else {
        throw new Error("File '" + path + "' not found!");
    }
};

Mplayer.prototype.getTimeLength = function(callback) {
    if(this.childProc !== null){
        this.rl.question("get_time_length\n", function(answer) {
            callback(answer.split('=')[1]);
        });
    }
};

Mplayer.prototype.getTimePosition = function(callback) {
    if(this.childProc !== null){
        this.rl.question("get_time_pos\n", function(answer) {
            callback(answer.split('=')[1]);
        });
    }
};

module.exports = new MPlayer();
